File: 0039-combination-sum/0039-combination-sum.kt
class Solution {
    fun combinationSum(candidates: IntArray, target: Int): List<List<Int>> {
        val results: MutableList<List<Int>> = mutableListOf()

        fun dfs(target: Int, index: Int, path: ArrayDeque<Int>) {
            if (target < 0) return
            if (target == 0) {
                results.add(ArrayList(path))
                return
            }

            for (i in index until candidates.size) {
                path.add(candidates[i])
                dfs(target - candidates[i], i, path)
                path.removeLast()
            }
        }

        dfs(target, 0, ArrayDeque())

        return results
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: list[int], target: int) -> list[list[int]]:
        result = []
        cur = []

        def backtrack(start, remaining: int):
            if remaining == 0:
                result.append(cur[:])
                return
            
            if remaining < 0:
                return
            
            for i in range(start, len(candidates)):
                cur.append(candidates[i])
                backtrack(i, remaining - candidates[i])
                cur.pop()
        
        backtrack(0, target)

        return result
        